File: src/main/java/com/nickcoblentz/montoya/MontoyaLogger.java
package com.nickcoblentz.montoya;

import burp.api.montoya.MontoyaApi;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

public class MontoyaLogger {

    private final MontoyaApi _api;
    private String _LogLevel;
    public final static String DebugLogLevel="DEBUG";
    public final static String WarnLogLevel="WARN";
    public final static String ErrorLogLevel="ERROR";
    public final static List<String> LogLevelOptions = List.of(DebugLogLevel,WarnLogLevel,ErrorLogLevel);
    private List<String> _selectedLogLevels = new LinkedList<String>();


    public MontoyaLogger(MontoyaApi api) {
        this._api = api;
        setDefaultLogLevel();
        //_api.logging().logToOutput("got here 1");

    }

    public MontoyaLogger(MontoyaApi api, String logLevel) {
        this._api = api;
        setLogLevel(logLevel);
        //.logging().logToOutput("got here 2: "+logLevel);
    }

    public void setDefaultLogLevel()
    {
        //_api.logging().logToOutput("got here 3: ");
        setLogLevel(this.ErrorLogLevel);
    }

    public void setLogLevel(String logLevel)
    {
//        _api.logging().logToOutput("got here 4: "+logLevel);
        _selectedLogLevels.clear();
        _selectedLogLevels.add(this.ErrorLogLevel);

        this._LogLevel = switch(logLevel) {
        case DebugLogLevel:
//            _api.logging().logToOutput("got here 4: debug");
            _selectedLogLevels.add(this.DebugLogLevel);
        case WarnLogLevel:
//            _api.logging().logToOutput("got here 4: warn");
            _selectedLogLevels.add(this.WarnLogLevel);
            yield logLevel;
        default:
//            _api.logging().logToOutput("got here 4: default");
            yield this.ErrorLogLevel;
        };

//        _api.logging().logToOutput("got here 4.5: "+_LogLevel);
    }

    public String getLogLevel()
    {
        return _LogLevel;
    }

    public String getClassAndMethodName()
    {
        StackWalker walker = StackWalker.getInstance();
        AtomicReference<String> className= new AtomicReference<>("");
        AtomicReference<String> methodName= new AtomicReference<>("");
        StackWalker.StackFrame sf = walker.walk((s) ->{
            return s.skip(3).findFirst().get();
        });
        className.set(sf.getClassName());
        methodName.set(sf.getMethodName());
        return className.get()+"::"+methodName.get();
    }
    public void debugLog(String message)
    {
        log(DebugLogLevel,message);
    }
    public void debugLog(String className, String message)
    {
        log(DebugLogLevel,className, message);
    }

    public void warnLog(String message)
    {
        log(WarnLogLevel,message);
    }
    public void warnLog(String className, String message)
    {
        log(WarnLogLevel,className,message);
    }
    public void errorLog(String message)
    {
        log(ErrorLogLevel,message);
    }

    public void errorLog(String className, String message)
    {
        log(ErrorLogLevel,className,message);
    }

    public void log(String logLevel,String message)
    {
        log(logLevel,getClassAndMethodName(),message);
    }
    public void log(String logLevel, String className, String message)
    {
        if(!shouldLog(logLevel))
            return;

        StringBuilder logBuilder = new StringBuilder();
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        Date date = new Date();
        logBuilder.append(String.format("%s\t|%s\t|%s - %s",logLevel,dateFormat.format(date),className,message));

//        _api.logging().logToOutput("got here 5?: "+logLevel);

        if(logLevel.equals(this.ErrorLogLevel))
        {
            _api.logging().logToError(logBuilder.toString());
        }
        else
        {
            _api.logging().logToOutput(logBuilder.toString());
        }
    }

    public boolean shouldLog(String logLevel)
    {
//        _api.logging().logToOutput("got here log?: "+this._selectedLogLevels.contains(logLevel));
        return this._selectedLogLevels.contains(logLevel);
    }
    
}

File: src/main/java/com/nickcoblentz/montoya/MontoyaLogger.kt
package com.nickcoblentz.montoya

import burp.api.montoya.MontoyaApi
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.atomic.AtomicReference
import java.util.stream.Stream

enum class LogLevel {
    DEBUG,
    WARN,
    ERROR
}

class MontoyaLogger {
    private val _api: MontoyaApi
    private var _currentLogLevel = LogLevel.ERROR

    constructor(api: MontoyaApi) {
        this._api = api

        //_api.logging().logToOutput("got here 1");
    }

    constructor(api: MontoyaApi, logLevel: LogLevel) {
        this._api = api
        _currentLogLevel = logLevel
        //.logging().logToOutput("got here 2: "+logLevel);
    }

    private fun classAndMethodName(): String
    {
        val walker = StackWalker.getInstance()
        val className = AtomicReference("")
        val methodName = AtomicReference("")
        val sf = walker.walk { s: Stream<StackWalker.StackFrame?> ->
            s.skip(
                3
            ).findFirst().get()
        }
        className.set(sf.className)
        methodName.set(sf.methodName)
        return className.get() + "::" + methodName.get()
    }

    fun debugLog(message: String?) {
        log(LogLevel.DEBUG, message)
    }

    fun debugLog(className: String?, message: String?) {
        log(LogLevel.DEBUG, className, message)
    }

    fun warnLog(message: String?) {
        log(LogLevel.WARN, message)
    }

    fun warnLog(className: String?, message: String?) {
        log(LogLevel.WARN, className, message)
    }

    fun errorLog(message: String?) {
        log(LogLevel.ERROR, message)
    }

    fun errorLog(className: String?, message: String?) {
        log(LogLevel.ERROR, className, message)
    }

    fun log(logLevel: LogLevel, message: String?) {
        log(logLevel, classAndMethodName(), message)
    }

    fun log(logLevel: LogLevel, className: String?, message: String?) {
        if (!shouldLog(logLevel)) return

        val logBuilder = StringBuilder()
        val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        val date = Date()
        logBuilder.append(String.format("%s\t|%s\t|%s - %s", logLevel, dateFormat.format(date), className, message))

        //        _api.logging().logToOutput("got here 5?: "+logLevel);
        when(logLevel) {
            LogLevel.ERROR -> _api.logging().logToError(logBuilder.toString())
            else ->_api.logging().logToOutput(logBuilder.toString())
        }
    }

    private fun shouldLog(logLevel: LogLevel): Boolean {
//        _api.logging().logToOutput("got here log?: "+this._selectedLogLevels.contains(logLevel));
        return when(logLevel) {
            _currentLogLevel -> true
            LogLevel.ERROR -> true
            LogLevel.WARN -> _currentLogLevel==LogLevel.DEBUG
            else -> false
        }
    }

}